### PR TITLE
Added a generic provider proof of concept

### PR DESCRIPTION
**Pull request by nvg-smile**
Added support for generic providers.

### CODE REVIEW

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/provider/MdmProviderLoader.java [hunk 1]
```diff
@@ -45,17 +47,30 @@
...
-				myResourceProviderFactory.addSupplier(() -> new MdmProviderDstu3(myFhirContext, myMdmControllerSvc, myMdmMatchFinderSvc, myMdmExpungeSvc, myMdmSubmitSvc));
+				myResourceProviderFactory.addSupplier(() -> {
+					myMdmProvider = new MdmProviderDstu3(myFhirContext, myMdmControllerSvc, myMdmMatchFinderSvc, myMdmExpungeSvc, myMdmSubmitSvc);
+					return myMdmProvider;
+				});
 				break;
 			case R4:
-				myResourceProviderFactory.addSupplier(() -> new MdmProviderR4(myFhirContext, myMdmControllerSvc, myMdmMatchFinderSvc, myMdmExpungeSvc, myMdmSubmitSvc));
+				myResourceProviderFactory.addSupplier(() -> {
+					myMdmProvider = new MdmProviderR4(myFhirContext, myMdmControllerSvc, myMdmMatchFinderSvc, myMdmExpungeSvc, myMdmSubmitSvc);
```
Comment (tadgh) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/provider/MdmProviderLoader.java:62]: Shouldn't we no longer have MdmProviderDStu3 and MdmProviderR4?

**Review by tadgh — APPROVED** [review]
Approved, but i think you have to delete the MdmProviderDstu3 and R4 now.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/BundleBuilder.java [hunk 2]
```diff
@@ -155,6 +219,57 @@ public IBaseBundle getBundle() {
...
+	 * 		Element type to create
+	 * @param theInitialValue
+	 * 		Optional initial value to be set on the new instance, in case of multiple values, only the first one will be
+	 * 	set.
+	 * @param <T>
+	 *    	Actual type of the parameterized primitive type interface
+	 * @return
+	 * 		Returns the newly created instance
+	 */
+	public <T> IPrimitiveType<T> newPrimitive(String theTypeName, T... theInitialValue) {
```
Comment (jamesagnew) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/util/BundleBuilder.java:258]: @nvg-smile This feels like misleading API to have it take a varargs but only ever look at the first element. I know it's documented but it still feels like a recipe for bugs. Can this be dropped to a single arg? Maybe you could create a separate helper method like `public static <T> T firstElementOrNull(T...)` whose output could be passed into this method to make what's going on a bit more explicit

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/BundleBuilder.java [hunk 3]
```diff
@@ -155,6 +219,57 @@ public IBaseBundle getBundle() {
...
+		BaseRuntimeElementDefinition primitiveDefinition = myContext.getElementDefinition(theTypeName);
+		if (primitiveDefinition == null) {
```
Comment (jamesagnew) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/util/BundleBuilder.java:260]: @nvg-smile Recommend replacing this null check with a single-line: 

```java
Validate.notNull(primitiveDefinition, "Unable to find definition for %s", theTypeName);
```

^ The above is shorter, and won't decrease your code coverage numbers

**Pull request closed without merging**